flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/materializedtable/MaterializedTableManager.java

```diff
@@ -804,6 +808,114 @@ protected static String getRefreshStatement(
         return insertStatement.toString();
     }
 
+    private ResultFetcher callAlterMaterializedTableAsQueryOperation(
+            OperationExecutor operationExecutor,
+            OperationHandle handle,
+            AlterMaterializedTableAsQueryOperation op) {
+        ObjectIdentifier tableIdentifier = op.getTableIdentifier();
+        CatalogMaterializedTable oldMaterializedTable =
+                getCatalogMaterializedTable(operationExecutor, tableIdentifier);
+
+        if (CatalogMaterializedTable.RefreshMode.FULL == oldMaterializedTable.getRefreshMode()) {
+            // direct apply the alter operation
+            List<TableChange> changes = new ArrayList<>(op.getColumnChanges());
+            changes.add(op.getDefinitionQueryChange());
+            AlterMaterializedTableChangeOperation alterMaterializedTableChangeOperation =
+                    new AlterMaterializedTableChangeOperation(
+                            tableIdentifier, changes, op.getNewMaterializedTable());
+            return operationExecutor.callExecutableOperation(
+                    handle, alterMaterializedTableChangeOperation);
+        }
+
+        // 1. suspend the materialized table
+        if (CatalogMaterializedTable.RefreshStatus.SUSPENDED
+                != oldMaterializedTable.getRefreshStatus()) {
+            suspendContinuousRefreshJob(
+                    operationExecutor, handle, tableIdentifier, oldMaterializedTable);
+        }
+
+        // 2. replace query definition and resume the materialized table
+        // alter materialized table schema & query definition & refresh handler
+        // we should reset the savepoint path after the alter operation
+        ContinuousRefreshHandler refreshHandler =
+                deserializeContinuousHandler(oldMaterializedTable.getSerializedRefreshHandler());
+        ContinuousRefreshHandler resetHandler =
+                new ContinuousRefreshHandler(
+                        refreshHandler.getExecutionTarget(), refreshHandler.getJobId());
+        byte[] serializedBytes = serializeContinuousHandler(resetHandler);
+
+        List<TableChange> tableChanges = new ArrayList<>(op.getColumnChanges());
+        tableChanges.add(op.getDefinitionQueryChange());
+        tableChanges.add(
+                TableChange.modifyRefreshHandler(resetHandler.asSummaryString(), serializedBytes));
+        CatalogMaterializedTable updatedMaterializedTable =
+                op.getNewMaterializedTable()
+                        .copy(
+                                op.getNewMaterializedTable().getRefreshStatus(),
```

Comment:
You should use the refresh status and refresh handler that you get when you resume the refresh job.